text/0061-minimal-streams.md

```diff
@@ -0,0 +1,275 @@
+
+<img src="./0061-minimal-streams/subtyping.svg" width="550" alt="a diagram showing node 'payload+ready+valid' subtyping node 'payload+ready', node 'payload+valid', and node 'payload'; and both node 'payload+ready' and node 'payload+valid' subtyping node 'payload'">
+
+Thus, option (B4) is the only feasible one: it does not require interface adaptation and avoids complex inheritance hierarchies in favor of a one-class solution while at the same time capturing the complex and directional subtyping relationship through the use of `amaranth.lib.wiring.connect()`. Also, since `connect()` can connect multiple streams, it is possible to broadcast a stream (connect a single transmitter to multiple receivers) with no interconnect, provided that none of the receivers require backpressure.
+
+### C: Payload shape
+[payload-shape]: #payload-shape
+
+There are three options for the payload member that are naturally possible with `lib.wiring`:
+
+1. Unrestricted `payload` member: a shape (port) or a signature (interface) are all valid, with no restrictions on the interface.
+2. `payload` member restricted to a shape (port) or a signature (interface) with only `Out` members.
+3. `payload` member restricted to a shape (port) only.
+    - This is the option used in this proposal.
+
+Option (C1) is not viable as it allows bidirectional communication to happen through the `payload` member, in contradiction to the goals of this proposal. Among other reasons, it makes streams no longer possible to pass through an async FIFO for clock domain crossing, which drastically reduces their usefulness.
+
+Option (C2) does have the advantage of splitting the payload into individual signals, which are more visible in waveform diagrams and on module boundaries in Verilog. However, it has a serious drawback, which is that `amaranth.lib.wiring.connect()` will not connect a single member to multiple members. This means that streams can no longer be directly connected to the three signals made available by the existing FIFOs in `amaranth.lib.fifo`, requiring adaptation.
+
+Option (C3) affords the same flexibility in the payload as option (2), without requiring stream-specific FIFOs; instead, an existing FIFO with its `[rw]_ack`, `[rw]_rdy`, `[rw]_data` signals can be connected to a stream endpoint provided that the same signals are put into an `amaranth.lib.wiring.PureInterface`, as long as the `[rw]_data` signal has the right width.
+
+### D: Omitted features
+[omitted-features]: #omitted-features
+
+This proposal intentionally omits the following features:
+1. packetization (often implemented using `first` and/or `last` control signals)
```

Comment:
I don't think it's wrong to exclude this, but I do think that it is an important feature to get to quickly. Many use cases need packetizetion, and if it's not provided may invent their own version.

Reply:
An issue that I discovered is that there's no consensus on even the structure of trivial packets. Some people want `first` only, some `last` only, some both `first` and `last`. And this is before going into interaction of packetization with other features.

Reply:
Maybe the fact that AXI-Stream uses `last` is a strong case for using `last`. I also find it easier to deal with than `first`. If the protocol has `last`, then when you see it asserted, you usually must reset things for the start of the next packet, and you can do it sequentially. A `first`, instead, usually forces you to deal with this situation combinationally, or to stop processing for one cycle.

Reply:
> Maybe the fact that AXI-Stream uses `last` is a strong case for using `last`.

Not really. It is important to remain *interoperable* to some extent with existing protocols, yes, but the fact that AXI-Stream did something is not in itself a motivation for us doing something the same way. There are plenty of decisions in AMBA we should absolutely not replicate.

> If the protocol has `last`, then when you see it asserted, you usually must reset things for the start of the next packet, and you can do it sequentially. A `first`, instead, usually forces you to deal with this situation combinationally, or to stop processing for one cycle.

Other people (@zyp) argued for allowing `first` without `last` for cases where you don't know if you have a last byte, e.g. COBS. This is what I'm referring to as a lack of consensus on what the protocol should be.